ejb3/src/main/java/org/jboss/as/ejb3/tx/CMTTxInterceptor.java

```diff
@@ -79,7 +79,7 @@
      */
     protected void endTransaction(final TransactionManager tm, final Transaction tx) {
         try {
-            if (tx != tm.getTransaction()) {
+            if (! tx.equals(tm.getTransaction())) {
```

Comment:
Problem arises in following scenario:
1. Transaction is started.
2. Transaction reaper cancels the transaction.
3. CMTTxInterceptor#endTransaction is executed.

In more details what happens is: in point 1 transaction is associated with the thread in a normal way, but after it is cancelled there is still Current object associated with the thread but the transaction is closed. As a result if we are performing comparison and executing TransactionManager#getTransaction method we would be delegated to com.arjuna.ats.internal.jta.transaction.jts.TransactionImple#getTransaction but there will be no transaction in _transactions map and a new object will be created - it will represent the same transaction. This is not the only fix necessary - we need also narayana fix https://github.com/jbosstm/narayana/pull/1044 - the objects will be compared as equal - which is correct.

Regarding disassociating the transaction from the thread: my current understanding is that my first diagnosis of the problem is incorrect and that it works correctly: if transaction is recognized by CMTTxInterceptor as rolled back it will be disassociated from the thread in CMTTxInterceptor#endTransaction#97.

What is also interesting is that transaction which stayed associated with the thread incorrectly was not causing problem in most scenarios as it was removed as the side effect of suspend/resume invocations of another transactional call. The only problem was if the next invocation was not transactional - f.e. corbaloc invocation to NamingService - in such scenario there was no part of code to remove the transaction as a side effect the error ocurred. 

Anyway I believe that after this fix + narayana one the above scenario should work correctly.